test/cpp/api/serialize.cpp

```diff
@@ -501,25 +511,45 @@ TEST(SerializeTest, Optim_Adagrad) {
   is_optimizer_state_equal<AdagradParamState>(optim1.state(), optim1_2.state());
 }
 
-TEST(SerializeTest, SerializationShouldPreserveIteration_SGD) {
-  std::vector<torch::Tensor> parameters = {
-      torch::randn({2, 2}), torch::randn({3, 3})};
-
-  torch::optim::SGD optimizer(parameters, 1.0);
-
-  optimizer.step();
-  optimizer.step();
+TEST(SerializeTest, Optim_SGD) {
+  test_serialize_optimizer<SGD, SGDOptions, SGDParamState>(SGDOptions(1e-1).momentum(0.9));
 
-  ASSERT_EQ(optimizer.iteration(), 2);
-
-  auto tempfile = c10::make_tempfile();
-  torch::save(optimizer, tempfile.name);
+  // bc compatibility check
+  auto model1 = Linear(5, 2);
+  auto model1_params = model1->parameters();
+  // added a tensor for lazy init check - when all params do not have a momentum buffer entry
+  model1_params.emplace_back(torch::randn({2,3}));
+  auto optim1 = torch::optim::SGD(model1_params, torch::optim::SGDOptions(0.01).momentum(0.9));
 
-  torch::optim::SGD optimizer_out(parameters, 1.0);
-  ASSERT_EQ(optimizer_out.iteration(), 0);
+  auto x = torch::ones({10, 5});
+  auto step = [&x](torch::optim::Optimizer& optimizer, Linear model) {
+    optimizer.zero_grad();
+    auto y = model->forward(x).sum();
+    y.backward();
+    optimizer.step();
+  };
+  step(optim1, model1);
+  auto optim1_2 = SGD(model1_params, torch::optim::SGDOptions(1e-1));
 
-  torch::load(optimizer_out, tempfile.name);
-  ASSERT_EQ(optimizer_out.iteration(), 2);
+  std::vector<at::Tensor> momentum_buffers;
+  int64_t iteration_{0};
+  const auto& params_ = optim1.param_groups()[0].params();
+  const auto& optim1_state = optim1.state();
+  for (size_t i = 0; i < params_.size(); i++) {
+    if(i != (params_.size() - 1)) {
+      auto key_ = c10::guts::to_string(params_[i].unsafeGetTensorImpl());
+      const SGDParamState& curr_state_ = static_cast<const SGDParamState&>(*(optim1_state.at(key_).get()));
+      momentum_buffers.emplace_back(curr_state_.momentum_buffer());
+    }
+  }
```

Comment:
maybe we can assert `momentum_buffers.size() == (params_.size() - 1)` here, just to be extra safe

Reply:
before we deserialize?
I think it's clear we are only storing param_.size() - 1 elements in momentum_buffer and after we deserialize, we check if the state is equal where the state size for optimizer before and after deserialization is checked to be equal, so I don't think we need this check